crates/bevy_ecs/src/system/commands/mod.rs

```diff
@@ -837,30 +837,62 @@ impl<'w, 's, 'a> EntityCommands<'w, 's, 'a> {
         self.commands
     }
 
-    /// Allows querying of an entity's components.
+    /// Allows querying of an entity's components. This is most useful when working with a highly
+    /// composed entity where you wish to modify some of its components.
     ///
-    /// # Panics
-    ///
-    /// The command will panic when applied if the associated entity and/or queried components do not exist.
+    /// If the entity is not available or query fails, then the provided closure will not be executed.
     ///
     /// # Examples
     ///
+    /// ## Conditionally Modify A Component
     /// ```
     /// # use bevy_ecs::prelude::*;
     /// #
     /// # #[derive(Resource)]
     /// # struct PlayerEntity { entity: Entity }
+    /// #
     /// #[derive(Component)]
     /// struct Health(u32);
     ///
-    /// fn zero_health(mut commands: Commands, player: Res<PlayerEntity>) {
+    /// #[derive(Component)]
+    /// struct LuckyStart;
+    ///
+    /// fn spawn_player(mut commands: Commands, player: Res<PlayerEntity>) {
+    ///     commands
+    ///         .entity(player.entity)
+    ///         .insert((Health(100), LuckyStart))
+    ///         /* ... */
+    ///         .with_query::<(&mut Health, &LuckyStart)>(|(mut health, _)| {
+    ///             // Players with the LuckyStart component start with double health.
+    ///             health.0 *= 2;
+    ///         });
```

Comment:
This doesn't seem like an improvement on:
```rust
let player_commands = commands.entity(player.entity);
let base_health = 100;
if give_lucky_start {
    player_commands.insert((Health(2 * base_health), LuckyStart));
} else {
    player_commands.insert(Health(health));
}
```

A system that applies the health bonus when a `LuckyStart` is added seems like it would be best.
Having the rule applied in a system would make it more discoverable and it would apply universally to anything that gets a `LuckyStart` such as enemies etc.

Reply:
I've added an alternate resolution to #4917 [here](https://github.com/bevyengine/bevy/issues/4917#issuecomment-1692531837), which basically moves this PR into a user-defined `trait`. The goal of this PR was just to close #4917, which is over a year old now. If this functionality isn't desired, I think it would be sensible to close #4917 and this PR.